### PR TITLE
Add Hub-Pattern Documentation

### PR DESCRIPTION
**Pull request by Lex-ari**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**| n|
|**_Documentation Included (y/n)_**| y |

---
## Change Description

Add Detailed Hub-Pattern Documentation. This should help with future hub-pattern projects. 

## Testing/Review Recommendations

Review organization / overall setup of the detailed documentation. Will appreciate any suggestions including grammar / wording and adding a new specific page / reorganizing.

A github-pages has been generated [here](https://lex-ari.github.io/fprime/UsersGuide/best/hub-pattern.html)

Note: There is also an [example repository](https://github.com/Lex-ari/fprime-two-pretty-best-friends) that should be migrated to fprime-community before this PR is merged. A link reference to the repository at the bottom of the Hub documentation will need to be updated.

- [x] [fprime-community example linked](https://github.com/fprime-community/fprime-hub-pattern-example)

CC: @thomas-bc @Joshua-Anderson @kevin-f-ortega @timcanham

### CODE REVIEW

File: docs/UsersGuide/best/hub-pattern.md [hunk 1]
```diff
@@ -17,4 +17,134 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+    configureTopology();
+    regCommands();
+    ...
+} 
+
+```
+
+## Creating a Combined Dictionary for a Multi-deployment System
+
+Since this system is running on two deployments, the dictionaries would need to be merged together in order to process the data from both
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: > Since this system is running on two deployments, the dictionaries would need to be merged together in order to process the data from both  deployments.

What about something like:

Since this system is running on two deployments, the command, telemetry, and event dictionaries need to be merged together so the GDS can process data received from both deployments.

File: docs/UsersGuide/best/hub-pattern.md [hunk 2]
```diff
@@ -17,4 +17,134 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+These can be either ran in the terminal or executed with a shell script.
+
+```shell
+#!/bin/bash
+
+fpp-depend ../build-fprime-automatic-native/locs.fpp topology.fpp -g deps.txt
+tr '\n' ',' < deps.txt | sed 's/,$//' > deps-comma.txt
+fpp-to-dict -i `cat deps-comma.txt` topology.fpp
+```
+> The second line takes the file dependencies and converts it to a one-line comma-separated input for `fpp-to-dict`. This line maybe omitted,
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: > This line maybe omitted,

File: docs/UsersGuide/best/hub-pattern.md [hunk 3]
```diff
@@ -17,4 +17,134 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+#!/bin/bash
+
+fpp-depend ../build-fprime-automatic-native/locs.fpp topology.fpp -g deps.txt
+tr '\n' ',' < deps.txt | sed 's/,$//' > deps-comma.txt
+fpp-to-dict -i `cat deps-comma.txt` topology.fpp
+```
+> The second line takes the file dependencies and converts it to a one-line comma-separated input for `fpp-to-dict`. This line maybe omitted, 
+which `fpp-to-dict` would then generate dictionaries for the GDS and both nodes.
+
+## Reference
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: I might recommend moving the example link up near the start so folks don't miss the complete example

File: docs/UsersGuide/best/hub-pattern.md [hunk 4]
```diff
@@ -17,4 +17,134 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+
+A possible usage of the Hub pattern includes running F´ on multiple computers acting as one F´ system. This example could 
+serve as a base for a multi-computer system running F´.
+
+![Detailed Hub](../media/detailed_hub_example.png)
+
+In this example, a basic implementation of the [Communication Adapter Interface](../../Design/
+communication-adapter-interface.md) 
+facilitates the transfer of serialized data between the two nodes across two different deployments. This interface 
+may be reinforced by including a `Fw::ComStub` and `Fw::ComQueue`, which are excluded in this implementation for
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: Reinforced can have several meanings and I'm wondering if it could confuse folks. What about phrasing the sentence like this?

While they are excluded from this example for simplicity, `Fw::ComStub` and `Fw::ComQueue` components could be added to make the communication link between hubs more robust.

File: docs/UsersGuide/best/hub-pattern.md [hunk 5]
```diff
@@ -17,4 +17,134 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+communication-adapter-interface.md) 
+facilitates the transfer of serialized data between the two nodes across two different deployments. This interface 
+may be reinforced by including a `Fw::ComStub` and `Fw::ComQueue`, which are excluded in this implementation for 
+simplicity. The Hub uses the `Data Out` port to export serialized data to the remote hub, which receives them on 
+the `Data In` port. This connection works reflexively, allowing the two hubs to communicate and transmit data 
+between each other.
+
+A [Command Splitter](../api/c++/html/svc_cmd_splitter.html) splits incoming commands from the Ground Interface to either a local or
+remote [Command Dispatcher](../api/c++/html/svc_cmd_dispatcher.html). This is inserted between the Ground Interface and the local Command 
+Dispatcher. The command offset of the command splitter should match the commands on the external node. An offset of 0x10000 will route any
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: > This is inserted between the Ground Interface and the local Command 
Dispatcher.

What about combining this with the previous sentence?

A Command Splitter sits between the Ground Interface and a local Command Dispatcher and splits commands from either a local or remote command dispatcher.

Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: I might say `command opcode` instead of command - that way it explicitly refers to the numerical opcode

**Review by Joshua-Anderson — COMMENTED** [review]
This is awesome @Lex-ari - I just added some suggestions we can discuss.

File: docs/UsersGuide/best/hub-pattern.md [hunk 6]
```diff
@@ -16,5 +16,132 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
+    ...
+    connections hub {
+      a_hub.LogSend -> a_eventLogger.LogRecv
+      a_hub.TlmSend -> a_tlmSend.TlmRecv
+    } 
+    ...
+```
+
+### Implementation Notes
+All instances names should be unique across the entire F´ System. Instances with the same name may
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: What about something like: Telemetry channels and events are currently named based on the component instance name, so re-using the same instance name between multiple deployments can result in naming conflicts.

File: docs/UsersGuide/best/hub-pattern.md [hunk 7]
```diff
@@ -16,5 +16,132 @@ ARINC 653 Channels, High-speed hardware buses between nodes, and UARTs between n
...
-There is now a standard implementation of the hub pattern. The [GenericHub](../api/c++/html/svc_generic_hub_component.html) is an
-implementation of the hub pattern that passes through F´ ports and `Fw::Buffer`s.
\ No newline at end of file
+There is now a standard implementation of the hub pattern. The 
+[GenericHub](../api/c++/html/svc_generic_hub_component.html) is an implementation of the hub pattern 
+that passes through F´ ports and `Fw::Buffer`s.
+
+## Detailed Example Generic Hub Setup
+
+A sample F´ project featuring a two-deployment hub setup can be found on [fprime-community](@FIXME).
```
Comment (Joshua-Anderson) [inline docs/UsersGuide/best/hub-pattern.md]: I might emphasize the example even more - maybe it's own section in the document?

**Review by Joshua-Anderson — COMMENTED** [review]
Looking really good!

**Review by Joshua-Anderson — APPROVED** [review]
Looks good to me - thanks @Lex-ari! @thomas-bc or @LeStarch should also take a look through before we merge.

**Review by thomas-bc — APPROVED** [review]
This is really nice! Could you fix the spelling errors? `NodeADeployment` and `NodeBDeployment` are tripping. Either add those words to `.github/actions/spelling/expect.txt`, or maybe reword to `Node_A_Deployment` or something which could be easier to read.

**Review by LeStarch — APPROVED** [review]
